### PR TITLE
generates loopy kernels

### PR DESCRIPTION
**Pull request by tj-sun**
This PR adds the capability to generate loopy kernels instead of coffee kernels.

I duplicated a lot of code in `kernel_interface/firedrake_loopy.py` because I think it will replace `kernel_interface/firedrake.py` eventually, when loopy supports all operations we need in firedrake, so probably not worthwhile to build a common parent class for the two interfaces.

The default behaviour of calling `compile_form()` is still generating a coffee kernel, so it shouldn't affect customers who uses tsfc other than firedrake.

### CODE REVIEW

File: tsfc/loopy.py [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+    if not domains:
+        domains = [isl.BasicSet("[] -> {[]}")]
+
+    # Create loopy kernel
+    knl = lp.make_kernel(domains, instructions, data, name=kernel_name, target=lp.CTarget(), seq_dependencies=True)
+
+    # Prevent loopy interchange by loopy
+    knl = lp.prioritize_loops(knl, ",".join(ctx.index_extent.keys()))
+
+    # Help loopy in scheduling by assigning priority to instructions
```
Comment (wence-) [inline tsfc/loopy.py]: Why not prioritise before making the kernel?

File: tsfc/loopy.py [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+        rank = []
+        for index in multiindex:
+            if isinstance(index, gem.Index):
+                rank.append(self.active_indices[index])
+            elif isinstance(index, gem.VariableIndex):
+                rank.append(expression(index.expression, self))
+            else:
+                assert isinstance(index, int)
+                rank.append(index)
+        return tuple(rank)
```
Comment (wence-) [inline tsfc/loopy.py]: rank is definitely the wrong noun, why not "indices"?

File: tsfc/loopy.py [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+
+    def pymbolic_variable(self, node):
+        try:
+            pym = self.gem_to_pymbolic[node]
+        except KeyError:
+            name = self.name_gen(node.name)
+            pym = p.Variable(name)
+            self.gem_to_pymbolic[node] = pym
+        if node in self.indices:
```
Comment (wence-) [inline tsfc/loopy.py]: How is a `Variable` ever a `Subscript`?  This is an odd smell.

Comment (tj-sun) [reply tsfc/loopy.py]: So if an gem node carries indices (as specified by `ImperoC.indices`), we return a `Subscript`? This is mimicing `tsfc.coffee._ref_symbol()`.

File: tsfc/loopy.py (hunk 2 above)
Comment (tj-sun) [reply tsfc/loopy.py]: Amended.

File: tsfc/loopy.py (hunk 1 above)
Comment (tj-sun) [reply tsfc/loopy.py]: Mainly that I don't know how many instructions I have in total, and it's nicer to have priorities go down from `len(instructions)` down to 0?

File: tsfc/kernel_interface/firedrake_loopy.py [hunk 4]
```diff
@@ -0,0 +1,356 @@
+from __future__ import absolute_import, print_function, division
```
[the remark this answers is not in the document]
Comment (tj-sun) [reply tsfc/kernel_interface/firedrake_loopy.py]: Amended.

File: tsfc/coffee.py [hunk 5]
```diff
@@ -83,12 +76,7 @@ def _ref_symbol(expr, parameters):
 
 
 def _root_pragma(expr, parameters):
-    """Decides whether to annonate the expression with
-    #pragma coffee expression"""
-    if expr in parameters.roots:
-        return "#pragma coffee expression"
-    else:
-        return None
+    return None
```
[the remark this answers is not in the document]
Comment (tj-sun) [reply tsfc/coffee.py]: Done.

File: tsfc/kernel_interface/firedrake_loopy.py [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+
+    def get_loopy_arguments(self):
+        args = [self.local_tensor, self.coordinates_arg]
+        if self.kernel.oriented:
+            args.append(self.cell_orientations_loopy_arg)
+        args.extend(self.coefficient_args)
+        if self.kernel.integral_type in ["exterior_facet", "exterior_facet_vert"]:
+            args.append(lp.GlobalArg("facet", dtype=numpy.uint32, shape=(1,)))
+        elif self.kernel.integral_type in ["interior_facet", "interior_facet_vert"]:
+            args.append(lp.GlobalArg("facet", dtype=numpy.uint32, shape=(2,)))
```
Comment (wence-) [inline tsfc/kernel_interface/firedrake_loopy.py]: Duplicate code (the same stuff appears in `construct_kernel`).

File: tsfc/loopy.py (hunk 3 above)
Comment (wence-) [reply tsfc/loopy.py]: OK.

File: tsfc/loopy.py (hunk 1 above)
Comment (wence-) [reply tsfc/loopy.py]: OK.

File: tsfc/loopy.py [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+    """
+    raise AssertionError("cannot generate loopy from %s" % type(tree))
+
+
+@statement.register(imp.Block)
+def statement_block(tree, ctx):
+    statements = []
+    for child in tree.children:
+        statements.extend(statement(child, ctx))
+    return statements
```
Comment (wence-) [inline tsfc/loopy.py]: AKA `return list(statement(child, ctx) for child in tree.children)`

Comment (wence-) [reply tsfc/loopy.py]: Oh wait, you get a list of statements, so `list(chain(*statement(child, ctx) for child in tree.children))`

File: tsfc/loopy.py [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+
+
+@_expression.register(gem.Failure)
+def _expression_failure(expr, parameters):
+    raise expr.exception
+
+
+@_expression.register(gem.Product)
+def _expression_product(expr, ctx):
+    return p.Product(tuple([expression(c, ctx) for c in expr.children]))
```
Comment (wence-) [inline tsfc/loopy.py]: No need to cons a list, why not: `tuple(expression(c, ctx) for c in expr.children))`

File: tsfc/loopy.py [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+
+
+@_expression.register(gem.Sum)
+def _expression_sum(expr, ctx):
+    return p.Sum(tuple([expression(c, ctx) for c in expr.children]))
```
Comment (wence-) [inline tsfc/loopy.py]: Ditto.

File: tsfc/loopy.py [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+
+
+@_expression.register(gem.Division)
+def _expression_division(expr, ctx):
+    return p.Quotient(*[expression(c, ctx) for c in expr.children])
```
Comment (wence-) [inline tsfc/loopy.py]: Also here: `p.Quotient(*(...))` doesn't cons a list.

File: tsfc/loopy.py [hunk 11]
```diff
@@ -0,0 +1,338 @@
...
+
+
+@_expression.register(gem.Power)
+def _expression_power(expr, ctx):
+    base, exponent = expr.children
+    return p.Power(expression(base, ctx), expression(exponent, ctx))
```
Comment (wence-) [inline tsfc/loopy.py]: Why not write this one as `p.Power(*(expression(c, ctx) for c in expr.children))`?

File: tsfc/driver.py [hunk 12]
```diff
@@ -266,7 +268,7 @@ def compile_expression_at_points(expression, points, coordinates, parameters=Non
     expression = ufl_utils.preprocess_expression(expression)
 
     # Initialise kernel builder
-    builder = firedrake_interface.ExpressionKernelBuilder()
+    builder = firedrake_interface_coffee.ExpressionKernelBuilder()
```
Comment (wence-) [inline tsfc/driver.py]: Could we migrate this to loopy too?

File: tsfc/loopy.py (hunk 1 above)
Comment (tj-sun) [reply tsfc/loopy.py]: Probably need to change this after we do dependencies correctly. Loopy doesn't seem to guarantee deterministic code generation though, which will be a problem.

File: tsfc/kernel_interface/firedrake_loopy.py (hunk 6 above)
Comment (tj-sun) [reply tsfc/kernel_interface/firedrake_loopy.py]: I've forgotten why I need this function in the first place, deleted.

File: tsfc/loopy.py (hunk 7 above)
Comment (tj-sun) [reply tsfc/loopy.py]: should it be
`list(chain.from_iterable(statement(child, ctx) for child in tree.children))` ?

File: tsfc/loopy.py (hunk 8 above)
Comment (tj-sun) [reply tsfc/loopy.py]: Oh yes!

File: tsfc/loopy.py (hunk 7 above)
Comment (wence-) [reply tsfc/loopy.py]: Either that or `list(chain(*(statement(...) ...)))` ?

File: tsfc/loopy.py (hunk 11 above)
Comment (tj-sun) [reply tsfc/loopy.py]: Sure.

File: tsfc/driver.py (hunk 12 above)
Comment (tj-sun) [reply tsfc/driver.py]: Let me try.

File: tsfc/loopy.py (hunk 7 above)
Comment (tj-sun) [reply tsfc/loopy.py]: Yep that's better.

File: requirements-git.txt [hunk 13]
```diff
@@ -2,3 +2,4 @@ git+https://github.com/coneoproject/COFFEE#egg=COFFEE
 git+https://github.com/firedrakeproject/ufl.git#egg=ufl
 git+https://github.com/firedrakeproject/fiat.git#egg=fiat
 git+https://github.com/FInAT/FInAT.git#egg=finat
+--recursive git+https://github.com/firedrakeproject/loopy.git#egg=loopy
```
Comment (miklos1) [inline requirements-git.txt]: So this recursive here is just wrong... That option should be used when you want to pull another requirements file.

File: tsfc/loopy.py [hunk 14]
```diff
@@ -0,0 +1,340 @@
...
+            return p.Variable("j0")(expression(arg, ctx))
+        elif nu == gem.one:
+            return p.Variable("j1")(expression(arg, ctx))
+    if name == 'yn':
+        nu, arg = expr.children
+        if nu == gem.Zero():
+            return p.Variable("y0")(expression(arg, ctx))
+        elif nu == gem.one:
+            return p.Variable("y1")(expression(arg, ctx))
+    return p.Variable(name)(*[expression(c, ctx) for c in expr.children])
```
Comment (miklos1) [inline tsfc/loopy.py]: This function needs updating... see current `tsfc/coffee.py`.

Comment (tj-sun) [reply tsfc/loopy.py]: Done -  but need testing on the complex branch.

File: tsfc/driver.py [hunk 15]
```diff
@@ -84,7 +89,7 @@ def compile_integral(integral_data, form_data, prefix, parameters, interface):
         _.update(parameters)
         parameters = _
     if interface is None:
-        interface = firedrake_interface.KernelBuilder
+        interface = firedrake_interface_loopy
```
Comment (miklos1) [inline tsfc/driver.py]: Needs update, I guess...

**Comment by miklos1** [comment]
OK, I find this mostly fine. The one thing I find most ad-hoc and ugly, are the random conditionals here and there about whether we are generating coffee or loopy... So the question is whether the presence of two code generation systems meant to be a short-term temporary thing, or a longer term thing.

- In the former case, I think it is not worth investing time to improve code quality in this respect, since the COFFEE code generation will go away soon anyway, and with that the problem of ad-hoc conditionals also goes away.
- In the latter case, however, a refactoring around which code generation pipeline is being used is strongly suggested.

File: tsfc/driver.py (hunk 15 above)
Comment (tj-sun) [reply tsfc/driver.py]: corrected.

**Comment by tj-sun** [comment]
> OK, I find this mostly fine. The one thing I find most ad-hoc and ugly, are the random conditionals here and there about whether we are generating coffee or loopy... So the question is whether the presence of two code generation systems meant to be a short-term temporary thing, or a longer term thing.

It is intended to be short term - functionalities such as Slate still need to go through COFFEE codegen before we extend loopy to incorporate more linear algebra intrinsics, which is not too hard to do but requires some careful planning.

Another thing is the `coffee=False` flag predates the `interface` parameter, so maybe now we can remove the `coffee` flag.

File: tsfc/driver.py [hunk 16]
```diff
@@ -56,6 +57,10 @@ def compile_form(form, prefix="form", parameters=None, interface=None):
     logger.info(GREEN % "compute_form_data finished in %g seconds.", time.time() - cpu_time)
 
     kernels = []
+    if coffee:
```
Comment (wence-) [inline tsfc/driver.py]: This now ignores the incoming interface and should say:
```python
if interface is None:
   if coffee:
       ...
   else:
```

Comment (wence-) [reply tsfc/driver.py]: Although that then means that we handle `interface` is None in two places. So perhaps move this logic to `compile_integral`

File: tsfc/driver.py [hunk 17]
```diff
@@ -84,7 +89,7 @@ def compile_integral(integral_data, form_data, prefix, parameters, interface):
         _.update(parameters)
         parameters = _
     if interface is None:
-        interface = firedrake_interface.KernelBuilder
+        interface = firedrake_interface_loopy.KernelBuilder
```
Comment (wence-) [inline tsfc/driver.py]: i.e. pass `coffee=True/False` into compile_integral and handle the relevant cases here.

File: tsfc/driver.py [hunk 18]
```diff
@@ -291,7 +295,10 @@ def compile_expression_at_points(expression, points, coordinates, interface=None
 
     # Initialise kernel builder
     if interface is None:
-        interface = firedrake_interface.ExpressionKernelBuilder
+        if coffee:
+            interface = firedrake_interface_coffee.ExpressionKernelBuilder
+        else:
+            interface = firedrake_interface_loopy.ExpressionKernelBuilder
```
Comment (wence-) [inline tsfc/driver.py]: OK, this one is right.

File: tsfc/driver.py (hunk 16 above)
Comment (dham) [reply tsfc/driver.py]: I don't really understand this code. Maybe you had better do it.

File: tsfc/kernel_interface/firedrake.py [hunk 19]
```diff
@@ -171,20 +172,25 @@ def require_cell_orientations(self):
...
-    def construct_kernel(self, return_arg, body):
+    def construct_kernel(self, return_arg, impero_c, precision, index_names):
         """Constructs an :class:`ExpressionKernel`.
 
         :arg return_arg: COFFEE argument for the return value
         :arg body: function body (:class:`coffee.Block` node)
         :returns: :class:`ExpressionKernel` object
         """
+
+        from tsfc.coffee import generate as generate_coffee
```
Comment (wence-) [inline tsfc/kernel_interface/firedrake.py]: Already imported at top level.

File: tsfc/driver.py (hunk 18 above)
Comment (dham) [reply tsfc/driver.py]: this is the one I did by hand!

**Comment by wence-** [comment]
Let's review #189 instead, the diff is identical, but without the WIP commits everywhere.

**Pull request closed without merging**